openedx/features/subscriptions/models.py

```diff
@@ -0,0 +1,89 @@
+
+    SUBSCRIPTION_TYPE_CHOICES = (
+        (LIMITED_ACCESS, 'Limited Access', ),
+        (FULL_ACCESS_COURSES, 'Full Access (Courses)', ),
+        (FULL_ACCESS_TIME_PERIOD, 'Full Access (Time period)', ),
+        (LIFETIME_ACCESS, 'Lifetime Access', ),
+    )
+
+    subscription_type = models.CharField(choices=SUBSCRIPTION_TYPE_CHOICES, max_length=30)
+    number_of_courses = models.PositiveIntegerField(null=True, blank=True)
+    course_enrollments = models.ManyToManyField(CourseEnrollment, blank=True)
+
+    def __str__(self):
+        return "Subscription of type {subscription_type} with ID {subscription_id} for User {user_email}".format(
+            subscription_type=self.subscription_type,
+            subscription_id=self.subscription_id,
+            user_email=self.user.username,
+        )
+
+    @property
+    def is_active(self):
+        """
+        Get active status of a subscription.
+
+        Check if a subscription is active even if a user can no longer purchase a new course through it.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.expiration_date and self.number_of_courses:
+            return self.expiration_date >= date.today() and course_enrollments_count <= self.number_of_courses
+        elif self.expiration_date and not self.number_of_courses:
+            return self.expiration_date >= date.today()
+        elif self.number_of_courses and not self.expiration_date :
+            return course_enrollments_count <= self.number_of_courses
+
+        return True
+
+    @property
+    def is_valid(self):
+        """
+        Get validity status of a subscription.
+
+        Check if a user can enroll in a new course through this subscription.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.subscription_type == self.LIFETIME_ACCESS:
+            return True
+        elif self.subscription_type == self.FULL_ACCESS_COURSES:
+            return self.expiration_date >= date.today()
+        elif self.subscription_type == self.FULL_ACCESS_TIME_PERIOD:
+            return course_enrollments_count < self.number_of_courses
+        elif self.subscription_type == self.LIMITED_ACCESS:
+            return self.expiration_date >= date.today() and course_enrollments_count < self.number_of_courses
+
+    @classmethod
+    def get_valid_subscription(self, user_id=None):
+        """
+        Get valid subscription of a user.
+        """
+        filter_params = dict()
+        if user_id:
+            filter_params['user'] = user_id
+
+        user_subscriptions = self.objects.filter(**filter_params)
+        valid_user_subscriptions = filter(lambda subscription: subscription.is_valid, user_subscriptions)
+        valid_user_subscriptions_ids = (subscription.subscription_id for subscription in valid_user_subscriptions)
+        return user_subscriptions.filter(subscription_id__in=valid_user_subscriptions_ids)
```

Comment:
If everything is correctly working, this function should return only one valid subscription even if `filter` is being used. I have used `filter` just so that if something goes wrong, the api doesn't completely crash. If u suggest that I enforce a check on count here, that can be done.

Reply:
Also, unless we write an `APIView`, we have to provide a queryset to be able to serialize it, if we return only on subscription in this, it raises error. I have added a check and log however to ensure that if valid count is one, it logs it and returns empty result.

Reply:
In this case we need to add `Returns` section in the docstring.